### ISSUE TITLE
Knowledge Distillation template

### ISSUE THREAD

**Issue by Separius** [pytorch/torchtitan#2109]
Hi, I want to use torchtitan for knowledge distillation, what is the right way to do it? should I hold both models inside the main model? (then how can I exclude the teacher from being saved or .train()ed or exclude it from the optimizer) or is there a way to have two separate models (with parallelism handled correctly, especially PP)?

if I have to hold both models in the same Model, then will this be a right forward()? (for now the assumption that both models have the same number of layers is ok)

```python
    def forward(
        self,
        tokens: torch.Tensor,
        tokens_t: torch.Tensor,
        attention_masks: AttentionMasksType | None = None,
    ):
        """
        Perform a forward pass through the Transformer model.
        """
        # passthrough for nonexistent layers, allows easy configuration of pipeline parallel stages
        h = self.tok_embeddings(tokens) if self.tok_embeddings else tokens
        h_t = self.tok_embeddings_t(tokens_t) if self.tok_embeddings_t else tokens_t

        for layer, layer_t in zip(self.layers.values(), self.layers_t.values()):
            h = layer(h, self.freqs_cis, attention_masks=attention_masks)
            h_t = layer(h_t, self.freqs_cis_t, attention_masks=attention_masks)
        h = self.norm(h) if self.norm else h
        h_t = self.norm(h_t) if self.norm_t else h_t
        output = self.output(h) if self.output else h
        output_t = self.output_t(h_t) if self.output_t else h_t
        return output, output_t
```

**Comment by tianyu-l** [comment]
We don't support knowledge distillation out of box today, but I imagine you could modify the train script to make this possible:

Are you going to put both models on the same set of devices? If so you could define two separate models, potentially apply different parallelisms on the two models, run the teacher under eval mode and get signal, and run the student model under training mode consuming the signal.

You don't have to run the two models at the same time, or have the layers intertwined (like in your code snippet). Does it make sense?

**Comment by Separius** [comment]
Thank you for the quick reply; having two models makes much more sense, but I think I'm still confused about how PP should work in this case, if both the teacher and the student have the same PP, can I make sure that all the GPUs are utilized well? should I write my own scheduler?

device 0: F1, T1, F2, T2, F3, T3, F4, ...
device 1: --,  F1, T1, F2, T2, F3, T3, ...
device 2: --, --,   F1, T1, F2, T2, B1, ...
device 3: --, --,   --,  F1, T1, B1, F2, ...

legend: F$i$ is forward for microbatch $i$ with the student, T$i$ is forward on the teacher, and B$i$ is backward

I think the much simpler way for my specific problem would be to use TP4 for the teacher and PP4 for the student (I have a mixed quantized student and was not able to make TP work with it, that is why I'm using PP)

**Comment by tianyu-l** [comment]
Interesting problem.

> I think the much simpler way for my specific problem would be to use TP4 for the teacher and PP4 for the student (I have a mixed quantized student and was not able to make TP work with it, that is why I'm using PP)

This sounds reasonable to me.

> should I write my own scheduler?

This should be doable, see this example https://github.com/pytorch/torchtitan/blob/1168f9e4d58bbd91c07b08c382d1ca3ae4b2e02c/tests/integration_tests/features.py#L285

cc @H-Huang if you could share more.
